src/hooks/use-dropdown-menu.tsx

```diff
@@ -0,0 +1,151 @@
+// Imports
+import React, { useState, useRef, createRef, useEffect } from 'react';
+
+
+// A custom Hook that abstracts away the listeners/controls for dropdown menus
+export default (itemCount: number) => {
+   // Use state
+   const [isOpen, setIsOpen] = useState<boolean>(false);
+   const [returnFocusToButton, setReturnFocusToButton] = useState<boolean>(false);
+   const currentFocusIndex = useRef<number | null>(null);
+   const firstRun = useRef(true);
+   
+   
+   // Create refs
+   const buttonRef = useRef<HTMLButtonElement>(null);
+   const itemRefs = useRef([...Array(itemCount)].map(() => createRef<HTMLAnchorElement>()));
+   
+   
+   // Create type guard
+   const isKeyboardEvent = (e: React.KeyboardEvent | React.MouseEvent): e is React.KeyboardEvent => (e as React.KeyboardEvent).key !== undefined;
+   
+   
+   // Handle changes in the state of the menu
+   useEffect(() => {
+       // Stop if the ref currently points to a null value
+       if (!itemRefs.current[0].current || !buttonRef.current) {
+           return;
+       }
+
+
+       // Stop if this is the first fire of the Hook, and update the ref
+       if (firstRun.current) {
+           firstRun.current = false;
+           return;
+       }
+
+       
+       // If the menu is currently open focus on the first item in the menu, else return focus to the button if state indicates that is desired behavior
+       if (isOpen) {
+           itemRefs.current[0].current.focus();
+           moveFocus(0);
+       } else if (returnFocusToButton) {
+           buttonRef.current.focus();
+       }
+   }, [isOpen, returnFocusToButton]);
+
+
+   // Handles moving the focus between menu items
+   const moveFocus = (itemIndex: number) => {
+        currentFocusIndex.current = itemIndex
+        itemRefs.current[itemIndex].current?.focus();
+   }
+   
+
+   // Create a handler function for the button's clicks and keyboard events
+   const buttonListener = (e: React.KeyboardEvent | React.MouseEvent) => {
+       // Detect if event was a keyboard event or a mouse event
+       if (isKeyboardEvent(e)) {
+           const { key } = e;
+           
+           if (!(key === 'Tab' || key === 'Shift')) {
+               e.preventDefault();
+           }
+           if (key === 'Enter' || key === ' ') {
+               setIsOpen(true);
+               setReturnFocusToButton(false);
+           } else if (key === 'Tab') {
+               setIsOpen(false);
+               setReturnFocusToButton(false);
+           }
+       } else {
+           setIsOpen(!isOpen);
+           setReturnFocusToButton(true);
```

Comment:
Is this necessary? This `else` block is dedicated to handling clicks. Assuming you've just clicked the button, then, the browser should automatically handle moving the focus to the button.

Reply:
It is probably not necessary since we aren't preventing the default behavior, this can be confirmed through testing later.

Reply:
Okay, let's remove it then.

```suggestion
```